### ISSUE TITLE
RFE: Provide templates for creating new systemd unit files

### ISSUE THREAD

**Issue by danimo** [systemd/systemd#22826]
**Is your feature request related to a problem? Please describe.**

The best way for a user to create a new unit/service file is `systemctl edit --full --force my.service`. The rest is left to the user, who needs to look up even basic boiler plating for writing a standard unit file. As a result, this often does not happen ([annecdotic evidence](https://twitter.com/fluepke/status/1506199342369193985)).

**Describe the solution you'd like**

1. `systemctl edit --full --force my.service` should look up a template from a template folder, depending on the unit type, and provide it to the editor environment. The elegance of a template folder is that it can be customized by downstream with ease.
2.  Consider introducing `systemctl create` or `systemctl create-unit` as a shorthand for `systemctl edit --full --force`

**Describe alternatives you've considered**

Writing a wrapper script. But it's never there where and when it is needed, so it fails to solve the issue.

**The systemd version you checked that didn't have the feature you are asking for**

I checked `master`, and Lennart asked me to file this RFE, so I guess it's safe to assume it's not there?

**Comment by keszybz** [comment]
I think `systemctl edit --new` would be better than a completely new top-level verb.

I think this would be useful. We'd just need to add code to look in a template directory for new units, which shouldn't be too hard, and then start populating that directory.

**Comment by poettering** [comment]
Yeah, "--full --force" is really weird wording for creating a new file. --new makes a lot of sense to me.

**Comment by DaanDeMeyer** [comment]
The relevant source code to modify will be in src/systemctl/systemctl.h, src/systemctl/systemctl.c, src/systemctl/systemctl-edit.h and src/systemctl/sytemctl-edit.c.

As discussed above, you should add a `--new` option to systemctl. When this is specified, we'll look in a directory for template files (e.g. /usr/share/systemd/unit/template) for a template file, and pre-populate the contents of the new unit file with the contents from the template file. To select the correct template, --new could take a string argument that allows users to specify which template they want to use. If no string argument is given for --new, we could list the possible templates the user can choose from.

To add support for --new, you'll need to modify `systemctl_parse_argv()` in systemctl.c. You can add a new global variable (this isn't frowned upon for argument variables) `arg_new` that is populated when the --edit option is specified. The code to look up the template and use it to prepopulate the unit file should go somewhere in systemctl-edit.c.

This should be enough information to get started. Take a look at the source code and try some stuff out. If you come up with any more questions, you can ask them here.

**Comment by cerebro1** [comment]
> The relevant source code to modify will be in src/systemctl/systemctl.h, src/systemctl/systemctl.c, src/systemctl/systemctl-edit.h and src/systemctl/sytemctl-edit.c.
> 
> As discussed above, you should add a `--new` option to systemctl. When this is specified, we'll look in a directory for template files (e.g. /usr/share/systemd/unit/template) for a template file, and pre-populate the contents of the new unit file with the contents from the template file. To select the correct template, --new could take a string argument that allows users to specify which template they want to use. If no string argument is given for --new, we could list the possible templates the user can choose from.
> 
> To add support for --new, you'll need to modify `systemctl_parse_argv()` in systemctl.c. You can add a new global variable (this isn't frowned upon for argument variables) `arg_new` that is populated when the --edit option is specified. The code to look up the template and use it to prepopulate the unit file should go somewhere in systemctl-edit.c.
> 
> This should be enough information to get started. Take a look at the source code and try some stuff out. If you come up with any more questions, you can ask them here.

Thanks a lot. This is really descriptive. I will go through the details.

**Comment by cerebro1** [comment]
Hi,
 So i have tried to go through the above details and created the flow. Please have a look and let me know if i am in the right direction with the logic and other stuff like i have used variables in expected way. 
![WhatsApp Image 2022-04-12 at 3 33 44 PM](https://user-images.githubusercontent.com/10183232/162935579-43cc0965-c1ad-4d22-87a3-042896ab196f.jpeg)
![WhatsApp Image 2022-04-12 at 3 33 44 PM (1)](https://user-images.githubusercontent.com/10183232/162935624-e5e0ae75-d93a-4433-8d5a-d2bb21784107.jpeg)
![WhatsApp Image 2022-04-12 at 3 33 43 PM](https://user-images.githubusercontent.com/10183232/162935641-6abc49a9-9c09-436d-8257-768b2b9d8ed3.jpeg)

**Comment by DaanDeMeyer** [comment]
Looks reasonable (please don't use pictures next time though, hand-writing is hard to read). Only thing I'm not sure about is tying the template names closely to unit types. There might be multiple templates for a single unit. I'd just let --new take an optional free-form string and we'd look if a file of the same name exists in the template directory (or maybe only files with the ".template" file extension or a relevant unit type extension). This is just my opinion though, other maintainers might disagree.

Regardless, this looks like a good starting point, next step is to implement an initial version and put up a PR. Any additional design considerations can be discussed there.

**Comment by cerebro1** [comment]
> Looks reasonable (please don't use pictures next time though, hand-writing is hard to read). Only thing I'm not sure about is tying the template names closely to unit types. There might be multiple templates for a single unit. I'd just let --new take an optional free-form string and we'd look if a file of the same name exists in the template directory (or maybe only files with the ".template" file extension or a relevant unit type extension). This is just my opinion though, other maintainers might disagree.
> 
> Regardless, this looks like a good starting point, next step is to implement an initial version and put up a PR. Any additional design considerations can be discussed there.

One last query before I move ahead, the template files don't exist right now. I am supposed to create a directory with template files of various unit types. How will that go with systemctl? I mean in this repo where am I supposed to keep that template directory and is there any script or function that I can use that will place the template directory in /use/share/systemd/unit/ ? 
Am I thinking right here?

**Comment by DaanDeMeyer** [comment]
There's a templates directory in factory/ in the repository. You can add a subdirectory in there that could contain templates shipped by systemd itself. You'll need to write some meson glue to make sure these templates are then installed by meson.

**Comment by cerebro1** [comment]
So to ensure i ahve correctly understood what you all have commented, i have divided like @anitazha mentioned here https://github.com/systemd/systemd/issues/22192#issuecomment-1098495357

Expected : systemctl edit --new=<template_name> <unit_name>

- If unit_name already exists, return error
- If template_name does not exist, return error
- If above two cases are not true, copy template_name file to unit_name file at desired location in /etc/system....
- If no template_name is given by user, copy a default template to unit_name file [ Could not understand how logically i can implement listing the templates for user so considered better to open default in that case. But this is my thought, please give your input for improvement]

So what have i done so far : 
- added --new and it is optional_argument.
- Raising PR and then will keep working step by step on that.

**Comment by bluca** [comment]
just one minor comment, apart from that looks like it's on the right track:

> If above two cases are not true, copy template_name file to unit_name file at desired location in /etc/system....

also check for --runtime flag, in which case it should go to /run/systemd/system/ instead

**Comment by amit-tewari** [comment]
@cerebro1 , @bluca ,
added a PR adjusting `--edit` functionality. Please review.

**Comment by cerebro1** [comment]
Resumed working on this.
